src/main/java/org/jabref/gui/util/Theme.java

```diff
@@ -33,45 +46,164 @@
-                additionalCssToLoad = cssResource;
-                LOGGER.debug("Using css {}", path);
-            } else {
-                additionalCssToLoad = Optional.empty();
-                LOGGER.warn("Cannot load css {}", path);
+            LOGGER.debug("Theme is {}, additional css path is {}, url is {}",
+                    this.type, cssPath.orElse(null), cssUrl.orElse(null));
+        }
+
+        additionalCssToLoad().ifPresent(this::loadCssToMemory);
+    }
+
+    /**
+     * Creates a Path from a path String
+     *
+     * @param pathString the path string
+     * @return the path on the default file system, or empty if not valid for that file system (e.g. bad characters)
+     */
+    private static Optional<Path> cssStringToPath(String pathString) {
+        try {
+            return Optional.of(Path.of(pathString));
+        } catch (InvalidPathException e) {
+            LOGGER.warn("Cannot load additional css {} because it is an invalid path: {}", pathString, e.getLocalizedMessage());
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Creates a URL from a file system path. The scheme of the URL depends on the file system provider, but it will
+     * generally be
+     *
+     * @param path the file system path
+     * @return the URL for that file system path
+     */
+    private static URL cssPathToUrl(Path path) {
+        try {
+            return path.toUri().toURL();
+        } catch (MalformedURLException e) {
+            LOGGER.warn("Cannot load additional css url {} because it is a malformed url: {}", path, e.getLocalizedMessage());
+            return null;
+        }
+    }
+
+    /**
+     * Returns the additional CSS file or resource, after checking that it is accessible.
+     *
+     * Note that the file checks are immediately out of date, i.e. the CSS file could become unavailable between
+     * the check and attempts to use that file. The checks are just on a best-effort basis.
+     *
+     * @return an optional providing the URL of the CSS file/resource, or empty
+     */
+    private Optional<URL> additionalCssToLoad() {
+        // When we have a valid file system path, check that the CSS file is readable
+        if (cssPath.isPresent()) {
+            Path path = cssPath.get();
+
+            if (!Files.exists(path)) {
+                LOGGER.warn("Not loading additional css file {} because it could not be found", cssPath.get());
+                return Optional.empty();
+            }
+
+            if (Files.isDirectory(path)) {
+                LOGGER.warn("Not loading additional css file {} because it is a directory", cssPath.get());
+                return Optional.empty();
+            }
+        }
+
+        return cssUrl;
+    }
+
+    /**
+     * Creates a data-embedded URL from a file or resource URL
+     * @param url the URL of the resource to convert into a data: url
+     */
+    private void loadCssToMemory(URL url) {
```

Comment:
Did you noticed significant performance improvements with this method? I was hoping that JavaFX internally caches the css file anyway.

Reply:
I think it does internally cache the CSS for Scenes at least, since most things continue to render OK when the CSS file is deleted. Although, if the file is deleted, then some dialog boxes like the About page and the Preferences are not themed. However, previews work differently. Indeed, the Preview Viewer is the only consumer of the getter that exposes the additional stylesheet, everything else uses installCss(). The Preview Viewer uses WebEngine, and that has its own CSS handling quite different from Scene. This is where the bug was -- the preferences dialog would thrown an uncaught exception because it has a preview tab.

Now, because WebEngine works differently and had a bug in how it handled CSS files in jmod files, we already had that the dark theme loaded into memory as a base-64 encoded URL. There was a TODO to remove that when the bug in WebEngine is fixed. However, rather than treating it as a temporary workaround, I have made it useful. By encoding custom themes as well as the dark theme into a URL, the WebEngine is less prone to error when the CSS file is removed.

As the dark theme already did this, I figured it was safe to take this approach and perform OK. However, performance wasn't the main motivator.

Reply:
Thank for the explanation! Since this code is somewhat of a hack (I remember I copy&pasted it from some JavaFX method), I would like to remove it as soon as the jmod issue is fixed (which should be JavaFX 15 or 16). So could you please readd the comment that this should be removed in the future. Thanks

Reply:
TODO added back onto loadCssToMemory() with a lot of detail, to make sure we resolve it without breaking this fix